fabric-data-generation-api-v1/src/client/java/net/fabricmc/fabric/impl/datagen/client/SoundTypeBuilderImpl.java

```diff
@@ -0,0 +1,217 @@
+		public static final Codec<SoundType> CODEC = RecordCodecBuilder.create(instance -> instance.group(
+				Entry.CODEC.listOf().fieldOf("sounds").forGetter(SoundType::sounds),
+				SOUND_CATEGORY_CODEC.fieldOf("category").forGetter(SoundType::category),
+				Codec.STRING.optionalFieldOf("subtitle").forGetter(SoundType::subtitle)
+		).apply(instance, SoundType::new));
+	}
+
+	private record Entry(Identifier name, RegistrationType type, float volume, float pitch, int weight, int attenuationDistance, boolean stream, boolean preload) {
+		private static final Codec<Entry> MAP_CODEC = RecordCodecBuilder.create(instance -> instance.group(
+				Identifier.CODEC.fieldOf("name").forGetter(Entry::name),
+				RegistrationType.CODEC.optionalFieldOf("type", RegistrationType.FILE).forGetter(Entry::type),
+				Codec.FLOAT.optionalFieldOf("volume", 1F).forGetter(Entry::volume),
+				Codec.FLOAT.optionalFieldOf("pitch", 1F).forGetter(Entry::pitch),
+				Codec.INT.optionalFieldOf("weight", 1).forGetter(Entry::weight),
+				Codec.INT.optionalFieldOf("attenuation_distance", 16).forGetter(Entry::attenuationDistance),
+				Codec.BOOL.optionalFieldOf("stream", false).forGetter(Entry::stream),
+				Codec.BOOL.optionalFieldOf("preload", false).forGetter(Entry::preload)
+		).apply(instance, Entry::new));
+
+		private static final Codec<Entry> STRING_CODEC = Identifier.CODEC.xmap(
+				id -> new Entry(id, RegistrationType.FILE, 1F, 1F, 1, 16, false, false),
+				Entry::name
+		);
+		private static final Codec<Entry> CODEC = Codec.xor(STRING_CODEC, MAP_CODEC).xmap(Either::unwrap, sound -> {
+			if (sound.type() != RegistrationType.FILE
+					|| sound.volume() != 1F
+					|| sound.pitch() != 1F
+					|| sound.weight() != 1
+					|| sound.attenuationDistance() != 16
+					|| sound.stream()
+					|| sound.preload()) {
+				return Either.right(sound);
+			}
+
+			return Either.left(sound);
+		});
+	}
+
+	public static final class EntryBuilderImpl implements EntryBuilder {
+		private final Identifier id;
+		private final RegistrationType type;
+
+		private float volume = 1F;
+		private float pitch = 1F;
+		private int attenuationDistance = 16;
+		private int weight = 1;
+		private boolean stream = false;
+		private boolean preload = false;
+
+		private EntryBuilderImpl(RegistrationType type, Identifier id) {
+			this.type = type;
+			this.id = id;
+		}
+
+		public static EntryBuilder create(RegistrationType type, Identifier id) {
+			return new EntryBuilderImpl(type, id);
+		}
+
+		public static EntryBuilder ofFile(Identifier soundFile) {
+			Objects.requireNonNull(soundFile, "Sound file/event id must not be null.");
+
+			if (soundFile.getPath().indexOf('.') != -1) {
+				LOGGER.warn("Sound file \"" + soundFile + "\" should not have a file extension and may result in the sound event not playing.");
```

Comment:
Is there ever a case were this will be neded? Maybe just throw?

Reply:
I want to leave the possibility open in case someone is trying to, for example, use .wav files